staging/src/k8s.io/client-go/discovery/cached/disk/compactable_disk_cache.go

```diff
@@ -0,0 +1,236 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package disk
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/peterbourgon/diskv"
+	"k8s.io/klog/v2"
+)
+
+const oneDay = 24 * time.Hour
+
+// CompactableDiskCache implements the httpcache.Cache interface. It delegates to
+// diskv.Diskv for Cache functions (Get, Set, Delete), and occasionally removes
+// old cached files. The specified "cacheSize" is NOT a hard limit; the disk
+// cache can breach this threshold. When cache compaction occurs, it will resize
+// the disk cache below this "cacheSize".
+//
+// Trigger: disk cache compactions are triggered when the modification time
+//   of an accessed cache file is older than one day. In this case, the
+//   accessed file's modification time is updated to the current time
+//   (file touched), and the cache compaction is run in parallel. Access
+//   time is not used directly because it is not well supported by all OSes.
+//
+// Concurrency:
+//   Within the same process: the goroutines of a single process are
+//   coordinated using "sync/singleflight" functionality.
+//
+//   Among different processes: if (for example) a separate kubectl command
+//   compacts the disk cache at the same time, separate processes can compete
+//   to delete the same files. This code must be tolerant to attempting to
+//   delete files already deleted.
+//
+type CompactableDiskCache struct {
+	cacheDir  string
+	cacheSize int64
+	diskv     *diskv.Diskv
+}
+
+// Get returns the []byte representation of a cached response and a bool
+// set to true if the value isn't empty. This function can trigger a
+// disk cache compaction which runs in a separate goroutine.
+func (tds *CompactableDiskCache) Get(key string) (responseBytes []byte, ok bool) {
+	filename := keyToFilename(key)
+	go possiblyCompactDiskCache(tds.cacheDir, filename, tds.cacheSize)
+	resp, err := tds.diskv.Read(filename)
+	if err != nil {
+		return []byte{}, false
+	}
+	return resp, true
+}
+
+// Set stores the []byte representation of a response against a key.
+func (tds *CompactableDiskCache) Set(key string, responseBytes []byte) {
+	key = keyToFilename(key)
+	tds.diskv.WriteStream(key, bytes.NewReader(responseBytes), true)
```

Comment:
FWIW the `true` here forces an fsync for every write. This causes modern Macs to take 10+ seconds to write out ~7MB of HTTP cache. Disabling it reduces that to about a second. See also https://github.com/kubernetes/kubernetes/pull/110851.